src/python/pants/engine/target.py

```diff
@@ -1387,7 +1377,109 @@ async def hydrate_sources(
         )
     )
     sources_field.validate_snapshot(snapshot)
-    return HydratedSources(snapshot, sources_field.filespec)
+
+    should_generate_sources = request.codegen_language is not None and isinstance(
```

Comment:
I think that rather than using the marker type for this,`HydrateSourcesRequest` should maybe have an API like the following:
```
class HydrateSourcesRequest:
    input: Sources
    output: Type[Sources] = Sources
```
ie: `HydrateSourcesRequest` would always have an output type to hydrate as, and the default output type would be `Sources`.

The output type would act as an assertion that the field was either: 1) already that type, 2) convertible to that type via codegen. The reason I think that this is important is that callers shouldn't need to be aware of the existence of codegen: so to the extent that it's possible, codegen should be invisible in the API other than in the rule that implements source hydration.

Reply:
This runs into some issues:

1) The `output` is a misnomer. We are not returning `Sources` or `PythonSources` etc. We are always returning `HydratedSources` no matter what.
2) Doesn't work well with "aggregator" rules like we have in `determine_source_files.py`. Very often, we will want to allow multiple distinct `Sources` classes; for example, https://github.com/pantsbuild/pants/pull/9638 allows `files()`, `resources()`, and Python targets. Asserting that `type(input) == output` means that the aggregator rule couldn't ergonomically support multiple disjoint `Sources` types like it currently does.

https://github.com/pantsbuild/pants/blob/fdc097576a0ef0dd1d3eeb0ebb535bf1b2da3887/src/python/pants/core/util_rules/determine_source_files.py#L35-L38

3. Some call sites explicitly do not want to use codegen, e.g. `filedeps`. Setting `HydrateSourcesRequest(sources_field, codegen_language=None)` expresses this much more clearly than `HydrateSourcesRequest(sources_field, output=Sources)` and relying on the implicit knowledge that there should be no code generators that go from `ThriftSources -> Sources` (which isn't ever enforced).

Reply:
> The output is a misnomer. We are not returning Sources or PythonSources etc. We are always returning HydratedSources no matter what.

The same goes for the codegen_language field. But the assumption is that the sources that end up captured are in fact sources that "qualify as that type".

> Very often, we will want to allow multiple distinct Sources classes; for example, #9638 allows files(), resources(), and Python targets. Asserting that type(input) == output means that the aggregator rule couldn't ergonomically support multiple disjoint Sources types like it currently does.

Somewhere, some caller needs to identify those three types and independently request them in order to operate on them. One approach might be to allow the caller to specify all of the output types that they could handle... then the callsite would look like:
```
sources_field = await Get[HydratedSources](HydrateSourcesRequest(sources_field, outputs=[Resources, Files, PythonSources]))
if isinstance(sources_field.value, Resources):
  ..
elif isinstance(..):
 ..
else:
 ..
```

> expresses this much more clearly than HydrateSourcesRequest(sources_field, output=Sources)

That would be the default for the field, so you would just specify `HydrateSourcesRequest(sources_field)`.